### DISCUSSION TITLE
Website Error Monitoring

### DISCUSSION THREAD

**Discussion by natankeddem** [zauberzeug/nicegui#1633] [Ideas / Feature Requests]
I had one additional comment in relation to https://github.com/zauberzeug/nicegui/pull/1617 and how you can find bugs like this more quickly. Is there any mechanism in your page hosting that would catch errors and alert? This bug did show errors in the console when it happened so I wonder if you can do anything to utilize that; if you aren't already. Something as simple as logging errors to a text file and having someone check it once a week. You could make the logging based on environmental variables so they wouldn't fire off in "non production" usage.

I was thinking anything you do here might give others ideas of how they can monitor website status and find any bugs in there own production websites more efficiently.

**Comment by rodja** [comment]
We use fly.io for hosting the page which already provides a way to view the logs. And with [Logs Shipper](https://fly.io/docs/going-to-production/monitoring/exporting-logs/) they can be persisted. It would indeed be beneficial to have a small program which receives the logs and makes them accessable/filterable. This could itself be a NiceGUI example!

**Comment by natankeddem** [comment]
What do you think about having a routable page on the existing nicegui.io site that might contain logging data? I am not sure if that opens up any security or other concerns. I could see why you guys would want to not expose that data.

**Comment by natankeddem** [comment]
It would be nice if any examples could be usable in a generic context without utilizing the built in logging helpers on fly.io. The Log Shipper feature sounds really well thought out and appears to have support for numerous endpoints, but I am not sure how many people utilize fly.io. For people self hosting or maybe using On Air another solution like this might be a starting point. Here is a rough POC for showing the console without any reliance on anything specific to fly.io. For more robust logging there would of course be usage of the `logging` module.

```py
import sys
from datetime import datetime
from io import StringIO
from nicegui import ui

output = StringIO()
sys.stdout = output
log=ui.log().classes('w-full h-96')

def show_logs():
    log.push(output.getvalue())
    output.truncate(0)

ui.timer(2.0, lambda: print(f'{datetime.now():%X}: This could be an error!'))
ui.timer(5.0, show_logs)

ui.run()
```

**Comment by rodja** [comment]
Interesting idea @natankeddem. I love the idea to expose the logs in general. It would enable the community to help us identifying possible bugs quickly and I'm general in favour of more "openness" and transparency. Of course we would need to check the security risks.
The main difficulty with your `sys.stdout` approach will be the adaptation into a multi-instance solution. We run between 14 and 30 servers world wide to provide short round-trip times for all users.

**Comment by natankeddem** [comment]
We could start off with something that at minimum lets the user know an exception has been thrown with a notification. When I first noticed that problem I thought it was a connectivity issue or something else going on with the website hosting or my ISP. It would have been helpful to see what exceptions were being thrown on the backend. Adding an exception catcher utilzing `ui.notify` might be just the thing with fairly minimal effort.

This won't allow your guys internally to track things without user intervention but it sounds like fly.io provides some of those tools, but there might need to be some effort to implement them per your earlier comment.

```py
import traceback
from nicegui import app, ui

ui.html('<style>.multi-line-notification { white-space: pre-line; }</style>')
app.on_exception(lambda e: ui.notify(
    'Website Exception: \n'
    f'{traceback.format_exception(e)} \n'
    'Please report to https://github.com/zauberzeug/nicegui/issues',
    multi_line=True,
    classes='multi-line-notification',))

def raise_exception():
    raise Exception

ui.button('Raise Exception', on_click=raise_exception)

ui.run()

```
